src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts

```diff
@@ -1857,6 +1889,111 @@ export class ConfigureWorkspaceFolderRecommendedExtensionsAction extends Abstrac
+		@INotificationService private notificationService: INotificationService,
+	) {
+		super(
+			'extensions.add-to-workspace-recommendations',
+			AddToWorkspaceRecommendationsAction.AddLabel,
+			contextService,
+			fileService,
+			editorService,
+			jsonEditingService,
+			textModelResolverService
+		);
+		this.class = AddToWorkspaceRecommendationsAction.AddClass;
+		this.enabled = false;
+		this.disposables.push(this.contextService.onDidChangeWorkbenchState(() => this.update()));
+		this.update();
+	}
+
+	private update(): void {
+		if (
+			this.contextService.getWorkbenchState() === WorkbenchState.EMPTY ||
+			!this.extension ||
+			this.extension.state !== ExtensionState.Installed
+		) {
+			this.enabled = false;
+		} else {
+			const folders = this.contextService.getWorkspace().folders;
+			const configurationFiles = folders
+				.map(workspaceFolder => workspaceFolder.toResource(paths.join('.vscode', 'extensions.json')));
+			const getFoldersRecommendedExtensionsPromises = configurationFiles.map(configurationFile => {
+				return this.getFolderRecommendedExtensions(configurationFile);
+			});
+			Promise.all(getFoldersRecommendedExtensionsPromises).then(foldersRecommendationsArray => {
+				// Don't enable the button if the selected extension is
+				// recommended in ANY the folders of the current workspace
+				this.enabled = !foldersRecommendationsArray.some((recommendations: string[]) => {
+					return recommendations.some(r => r === this.extension.id);
+				});
+				if (this.class === AddToWorkspaceRecommendationsAction.AddingClass) {
+					this.label = this.enabled ? AddToWorkspaceRecommendationsAction.AddLabel : AddToWorkspaceRecommendationsAction.AddedLabel;
+					this.class = this.enabled ? AddToWorkspaceRecommendationsAction.AddClass : AddToWorkspaceRecommendationsAction.AddedClass;
+				}
+			});
+		}
+	}
+
+	run(): TPromise<any> {
+		if (!this.enabled) {
+			return null;
+		}
+
+		this.enabled = false;
+		this.label = AddToWorkspaceRecommendationsAction.AddingLabel;
+		this.class = AddToWorkspaceRecommendationsAction.AddingClass;
+		const folders = this.contextService.getWorkspace().folders;
+		const pickFolderPromise = folders.length === 1
+			? TPromise.as(folders[0])
+			: this.commandService.executeCommand<IWorkspaceFolder>(PICK_WORKSPACE_FOLDER_COMMAND_ID);
+		return pickFolderPromise
+			.then(async workspaceFolder => {
+				if (workspaceFolder) {
+					const configurationFile = workspaceFolder.toResource(paths.join('.vscode', 'extensions.json'));
+					await this.addRecommendedExtensionToFolder(configurationFile, this.extension.id);
+					return false;
+				}
+				return true;
+			})
+			.then((pickCancelled: boolean) => {
+				if (!pickCancelled) {
+					this.notificationService.info(localize('AddToWorkspaceRecommendations.success', 'The extension was recommended in the selected folder.'));
+				}
+			}, err => {
+				this.notificationService.error(err);
+			})
+			.then(() => this.update());
```

Comment:
At this point, I am guessing all we need to do is update the label/class and enabled state of the action. If so, why not update them directly here? Calling `this.update()` would mean the `extensions.json` files of each folder will be read one more time for no reason.

Reply:
Yeah I was thinking the same thing actually, I'm also going to update the class and labeling back to the clickable state if the quickPick is cancelled, it would end up looking like this:

```ts
.then((pickCancelled: boolean) => {
	if (pickCancelled) {
		this.enableButton();
	} else {
		this.notificationService.info(localize('AddToWorkspaceRecommendations.success', 'The extension was recommended in the selected folder.'));
		this.enabled = false;
		this.label = AddToWorkspaceRecommendationsAction.AddedLabel;
		this.class = AddToWorkspaceRecommendationsAction.AddedClass;
	}
}, err => {
	this.notificationService.error(err);
	this.enableButton();
});
```

Look out for that change in the next commit